doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+```
+
+- You do not need to type annotate `self` or `cls`.
+- Do return `None` from a constructor.
+  
+```python
+# No:
+class KeyCredential:
+
+    def __init__(self):
+        ...  # I do not get type checked
+
+# Yes:
+class KeyCredential:
+
+    def __init__(self) -> None:
+        ...
+```
+
+- Do use [mypy](https://mypy.readthedocs.io/en/stable/) and [pyright](https://github.com/microsoft/pyright) type checkers to statically type check your client library code.
+- Do use [black](https://pypi.org/project/black/) to format type annotations.
+- Do add type hints directly to the source code. If you think you need to use stub files, check with the architects. Note that with stub files, type checking will only work for _users_ of the stub, but won't type check the library code itself.
+- Do mark your client library package to distribute type hints according to [PEP 561](https://peps.python.org/pep-0561/).
+- Do use the latest typing features available. If not supported by older versions of Python, consider taking a dependency and importing from `typing-extensions`.
+
+```python
+# from typing import TypedDict Python >3.8
+from typing_extensions import TypedDict
+```
+
+### Importing types
+
+- Do use only publicly exposed client library types in type hints. 
+- Do import types from modules such as `typing`, `typing_extensions`, `collections`, and `collections.abc`(Note that indexing support for generic collection types from `collections.abc` is only supported on Python 3.9+).
+- Do not import regular type hints under a `typing.TYPE_CHECKING` block. You may use `TYPE_CHECKING` to fix a circular import or avoid importing a type only needed in type annotations that is otherwise costly to load at runtime.
+
+```python
+from typing import TYPE_CHECKING
+
+# No:
+if TYPE_CHECKING:
+    from typing import Union, TypeVar, Any
+
+
+# Yes:
+if TYPE_CHECKING:
+    from a import b  # avoiding a circular import
+    from expensive import c  # avoiding runtime costs
+```
+
+- Use `from __future__ import annotations` for forward declarations or when using built-in generic collection types, like `dict` or `list`.
+
+```python
+from __future__ import annotations  # must be first import in file
+
+class Triangle:
+
+    @classmethod
+    def from_shape(cls) -> Triangle: # Don't need to make Triangle a forward reference / string
+        ...
+
+    def get_points(self) -> list[float]:  # allows use of list instead of typing.List
+        ...
+```
+
+### Ignoring type checkers
+
+- Do not silence the type checker with `type: ignore` unless other options are exhausted. Consider first using `typing.cast` or refactoring the code.
+- If you must use a `type: ignore`, try to be specific in what error code you're ignoring.
+
+```python
+# type: ignore[misc]  # mypy ignores only the error code in brackets
+
+# pyright: ignore[reportPrivateUsage]  # pyright ignores only the error code in brackets
+
+# type: ignore  # all errors ignored by both mypy and pyright
+```
```

Comment:
This example might mislead people into thinking that these comments have a global effect. Typically they are placed on a specific line of code, e.g.
```py
class Foo(Any):  # type: ignore # Can't inherit from Any in some versions
    ...
```